staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -711,28 +774,44 @@ func (c *Cacher) startDispatching(event *watchCacheEvent) {
 	// gain from avoiding memory allocations is much bigger.
 	c.watchersBuffer = c.watchersBuffer[:0]
 
-	// Iterate over "allWatchers" no matter what the trigger function is.
-	for _, watcher := range c.watchers.allWatchers {
-		c.watchersBuffer = append(c.watchersBuffer, watcher)
-	}
-	if supported {
-		// Iterate over watchers interested in the given values of the trigger.
-		for _, triggerValue := range triggerValues {
-			for _, watcher := range c.watchers.valueWatchers[triggerValue] {
+	if event.Type == watch.Bookmark {
+		// Pop up watchers from startBucketID to endBucketID, ignore those already
+		// stopped as we do not delete the watcher when it is stopped,
+		for bucketID := startBucketID; bucketID < endBucketID; bucketID++ {
+			for _, watcher := range c.bookmarkWatchers.popWatchers(bucketID) {
+				if watcher.stopped {
```

Comment:
nit:
```
if !watcher.stopped {
   c.watcherBuffer = append(...)
}
```

Also, probably you should use isStopped function for thread-safety?

Reply:
seed an event to a stopped watcher is fine, we do not need a lock here.

Reply:
it's not fine, because stop() is closing c.input (and sending something to close channel panics)

Reply:
stop() will be delayed at the end of dispatching...

Reply:
It might be better to never close c.input. See https://github.com/kubernetes/kubernetes/pull/75474#discussion_r273733310

Reply:
Until very recently (like week ago), this was the only signal to shutdown the process() function.
I guess you may be right, but I would really like to avoid bigger changes into this one - let's leave it for later simplifications.